feat(planes): porting AEP-111 from AIP-111 (#85)
aep-dev/aep.dev: aep/general/0111/aep.md

> # Planes

Reviewer: I added a bunch of comments below, some just proofreading and others more about the content, but my big question is:  what guidance does this AEP actually contain?  It doesn't seem like there's anything actionable.  I don't know what lint rules I'd write, or when I'd refer to it as part of API design.  What API design problem is this AEP addressing?

Author: at Google, there was a large debate about the type of design guidance that would be needed to provide really high-quality declarative APIs (e.g. for Terraform or Kubernetes), at odds with more flexible API needs (generally because of existing precedent, but also sometimes a matter of convenience).

To begin to help address those, we started adding this notion of API planes, and attached various guidance to the management plane in the AIPs: https://google.aip.dev/search?q=management.

to be fair this aligned with some terminology primarily used at Google. Many APIs at least differentiate between a "control plane" and a "data plane": https://docs.aws.amazon.com/whitepapers/latest/aws-fault-isolation-boundaries/control-planes-and-data-planes.html.

I think it's important for us to add some clarity on the various planes of operations and how to differentiate them, given it is common nomenclature for many APIs (especially in the cloud).

I think the content here is actually pretty solid, but it was definitely a work in progress

Reviewer: That's all useful context...but it doesn't actually answer the question about what actionable guidance is in this AEP.

> Resources and methods on an API can be divided into the _plane_ that they

Author: SGTM!

reside or perform operations upon. For the context of APIs, the following

> planes are defined:

Author: yep good point, I'll try to elaborate.

> - Management plane: a uniform, resource-oriented API that primarily configures

Author: this is in contrast to a data plane API (e.g. SQL server the API is via SQL statements).

> - Management plane: a uniform, resource-oriented API that primarily configures
>   and allows retrieval of resources.

Reviewer: ```suggestion
- Management plane: contains uniform, resource-oriented APIs that primarily configure
  and allow retrieval of resources.
```

Reviewer: Corresponding grammatical change for data plane, if you agree.

> - Data plane: a heterogenous API (ideally resource-oriented) that reads and

Reviewer: Why "ideally"?

Author: cases where you can't adhere to the AEPs are explained in the data-plane section. Do you think I should re-iterate that here?

  write user data. Often connects to entities provisioned by the management
  plane, such as virtual machines.

> 

Reviewer: I made a bunch of local suggestions above, but on the whole, I'm confused by the point this is trying to make.  The management plane reads and configures resources; the data plan is for reading and writing user data (also "ideally" in the form of resources).

I'm not sure that someone with no context could easily categorize APIs into the right plane given these definitions.

Reviewer: ...I'm not even sure that I can.  Would creating a Facebook group be a management plane or data plane method?  The group name/description/etc. are UGC, but if individual posts in the group are part of the data plane then does that place group management in the management plane?

If these resources are all in the same API, and APIs as a whole are categorized as data plane or management plane, does that mean this API is in the data plane?

What if `CreateBlobStore` and `CreateBlob` are in the same API?  Why wouldn't they be?

Author: I think on the whole we probably should discuss these in-person - would be a lot easier.

But answering your questions:

Facebook group would be management plane. Planes are somewhat contextual, but you create a facebook group and users post in them. 

> If these resources are all in the same API, and APIs as a whole are categorized as data plane or management plane, does that mean this API is in the data plane?

I don't think APIs have to be categorized as completely data / management. I think this could be done on the *operation* level. WDYT? I can update the doc.

> What if CreateBlobStore and CreateBlob are in the same API? Why wouldn't they be?

Yep that's fine, see above.

The term "plane" was originally used in networking architecture. Although

> system and network architecture often defines additional planes (e.g. control

Reviewer: ```suggestion
system and network architecture often defines additional plane (e.g. control
```

If "system and network architecture" is short for "system architecture and network architecture", which I think it is?

Author: yep! That's correct. 

I'm going to remove this section though - I don't think it's particularly relevant information, although maybe context is helpful?

> plane or power planes), as the AEPs are focused on the interface, they are not

Reviewer: ```suggestion
plane or power plane), as the AEPs are focused on the interface, they are not
```

> plane or power planes), as the AEPs are focused on the interface, they are not

Author: I deleted the section, but ack - I'll try to make sure other sections are clearer.

defined in this AEP.

## Guidance

### Management Plane

Management resources and methods exist primarily to provision, configure, and

> audit the resources that the data plane interfaces with.

Reviewer: Does the data plane "interface" with resources?  Or does it contain APIs that contain resources?

Author: data plane operates on resources - I'll try to clarify, PTAL.

> As an example, the following are considered management resources for a cloud

Reviewer: ```suggestion
For example, the following are considered management resources for a cloud
```

Nit:  I think this is a more common phrasing.

provider:

- virtual machines
- virtual private networks
- virtual disks
- a blob store instance
- a project or account

### Data Plane

Methods on the data plane operate on user data in a variety of data formats,
and generally interface with a resource provisioned via a management plane API.
Examples of data plane methods include:

- writing and reading rows in a table
- pushing to or pulling from a message queue
- uploading blobs to or downloading blobs from a blob store instance

> Data plane APIs **may** be heterogenous across a larger API surface, due to

Reviewer: What does it mean for an API to be heterogenous?  Not entirely resource-oriented?

What is a "larger API surface"?  Larger than what?

Author: `across a larger API surface` doesn't seem helpful here - will remove.

I probably can be more precise that heterogeneity - I mean that there's more leeway in the requirements of data plane vs those for management APIs, up to being completely non-AEP compliant (e.g. SQL or a protocol for blob storage).

requirements including high throughput, low latency, or the need to adhere to
an existing interface specification (e.g. ANSI SQL).

- For convenience, resources and methods that operate on the data plane **may**
  expose themselves via resource-oriented management APIs. If so, those
  resources and methods **must** adhere to the requirements of the management

>   plane as specified in the other AEPs.

Reviewer: Which other AEPs do (or will) talk about the requirements of a management plane API for its corresponding data plane APIs?

Author: https://google.aip.dev/search?q=plane. Primarily management plane operations have additional requirements around CRUDL.
